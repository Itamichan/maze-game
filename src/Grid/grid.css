
.grid {
    border: 1px solid black;
    flex-basis: 33%;
    height: 300px;
    align-self: center;
    font-size: 48px;
    padding-top: 110px;
}

.grid-container {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-evenly;
    align-items: center;
}
