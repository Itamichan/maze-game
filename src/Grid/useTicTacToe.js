import {useEffect, useState} from "react";


const useTicTacToe = (startPlayer = "x") => {
    const [gameState, setGameState] = useState("aaaaaaaaa");
    const [currentPlayer, setCurrentPlayer] = useState(startPlayer);

    // useEffect(() => {
    //     console.log(new Date());
    //     setGameState("aaaaaaaax")
    // }, [])


    let winner = "";
    if (gameState[0] === gameState[1] && gameState[0] === gameState[2]) {
        winner = gameState[0];
    } else if (gameState[3] === gameState[4] && gameState[3] === gameState[5]) {
        winner = gameState[3];
    } else if (gameState[6] === gameState[7] && gameState[6] === gameState[8]) {
        winner = gameState[6];
    } else if (gameState[0] === gameState[3] && gameState[0] === gameState[6]) {
        winner = gameState[0];
    } else if (gameState[1] === gameState[4] && gameState[1] === gameState[7]) {
        winner = gameState[1];
    } else if (gameState[2] === gameState[5] && gameState[2] === gameState[8]) {
        winner = gameState[2];
    } else if (gameState[0] === gameState[4] && gameState[0] === gameState[8]) {
        winner = gameState[0];
    } else if (gameState[2] === gameState[4] && gameState[2] === gameState[6]) {
        winner = gameState[2];
    }

    if (winner === 'a') {
        winner = ""
    }

    const hasWinner = winner !== "";
    const isDraw = !gameState.includes("a") && !winner;

    const onFieldClick = (index) => {
        if (gameState.charAt(index) === "a" && !winner) {
            setGameState(prevState => {
                let newString = prevState;
                newString = newString.substr(0, index) + currentPlayer + newString.substr(index + 1);
                // setCurrentPlayer(currentPlayer === "x" ? "0" : "x");
                return newString

            })
        }
    };

    let resetGame = () => {
        if (hasWinner || isDraw) {
            setGameState("aaaaaaaaa");
            setCurrentPlayer("x")
        }
    };


    return {
        gameState,
        currentPlayer,
        winner,
        hasWinner,
        isDraw,
        onFieldClick,
        resetGame
    }

};

export default useTicTacToe