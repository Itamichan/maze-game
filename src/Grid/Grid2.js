import React from 'react';
import './grid.css'
import useTicTacToe from "./useTicTacToe";

const Grid = () => {

    const {gameState, currentPlayer, winner, hasWinner, isDraw, onFieldClick, resetGame} = useTicTacToe("o");
    const showButton = isDraw || winner;

    let playerTurnInfo = <div> It is {currentPlayer} turn!</div>;
    if (hasWinner) {
        playerTurnInfo = <div> The winner is {winner}</div>

    } else if (isDraw) {
        playerTurnInfo = <div> It is a draw</div>
    }

    const gridRepresentation = gameState.split("").map((e, idx) => {
        let cellValue = e;
        if (e === "a") {
            cellValue = "";
        }
        return <div className="grid" onClick={() => onFieldClick(idx)}>{cellValue}</div>
    });

    return (
        <div>
            <div className="grid-container">{gridRepresentation}</div>
            {playerTurnInfo}
            {showButton && <button onClick={resetGame}>Reset</button>}
        </div>
    )
};

export default Grid