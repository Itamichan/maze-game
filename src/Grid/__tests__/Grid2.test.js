import React from 'react';
import {mount, shallow} from 'enzyme';
import Grid from "../Grid2";
import useTicTacToe from "../useTicTacToe";


describe('<Grid2/>', () => {

    it('should return the div with the winner', function () {
        const comp = shallow(<Grid/>);
        console.log(comp.debug());
    });

});

// define empty mock function
const fnClick = jest.fn();
describe('click events', () => {
    it('button click should show menu', () => {
        //replace actual function with mock function
        const component = shallow(<button onClick={fnClick}/>);

        //simulate a click
        component
            .find('button')
            .simulate('click');

        //check if function was called
        expect(fnClick).toHaveBeenCalled();
    });
});

describe('<Grid2/>', () => {

    it('should check that the div is called', function () {
        const comp = mount(<Grid/>);
        comp.find('div.grid').first()
            .simulate('click');

        console.log(comp.debug());

        expect(comp.debug()).toMatchSnapshot();

    });

});

// const myMock = jest.fn();
// console.log(myMock('return this string'));
// // > return this string
//
// // return `true` for first call,
// // return `false` for the second call
// // return a string 'hello mock' for the third call
// myMock.mockReturnValueOnce(true)
//     .mockReturnValueOnce(false)
//     .mockReturnValueOnce('hello mock');
// //call the mock three times to witness the results:
// console.log(myMock(), myMock(), myMock());