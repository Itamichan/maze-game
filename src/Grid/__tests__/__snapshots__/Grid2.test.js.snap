// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid2/> should check that the div is called 1`] = `
"<Grid>
  <div>
    <div className=\\"grid-container\\">
      <div className=\\"grid\\" onClick={[Function: onClick]}>
        o
      </div>
      <div className=\\"grid\\" onClick={[Function: onClick]} />
      <div className=\\"grid\\" onClick={[Function: onClick]} />
      <div className=\\"grid\\" onClick={[Function: onClick]} />
      <div className=\\"grid\\" onClick={[Function: onClick]} />
      <div className=\\"grid\\" onClick={[Function: onClick]} />
      <div className=\\"grid\\" onClick={[Function: onClick]} />
      <div className=\\"grid\\" onClick={[Function: onClick]} />
      <div className=\\"grid\\" onClick={[Function: onClick]} />
    </div>
    <div>
       It is 
      x
       turn!
    </div>
  </div>
</Grid>"
`;
